jdaviz/configs/mosviz/helper.py

```diff
@@ -51,86 +60,136 @@ def _extend_world(self, spec1d, ext):
+        # Workaround for flipped data
+        min_world, max_world = ((world[0], world[-1]) if not self._is_world_flipped()
+                                else (world[-1], world[0]))
+
+        # Warn the user if they've panned far enough away from the data
+        # that the viewers will desync
+        if min_1d < min_world or max_1d > max_world:
+            self._show_panning_warning()
+            return
+
+        self._panning_warning_triggered = False
+
+        idx_min = float((np.abs(world - min_1d)).argmin()) - extend_by
+        idx_max = float((np.abs(world - max_1d)).argmin()) - extend_by
+
+        self._update_in_progress = True
+        with self._scales2d.hold_sync():
+            self._scales2d.min = idx_min
+            self._scales2d.max = idx_max
+
+        self._update_in_progress = False
 
     def _update_spec1d_x_axis(self, change):
         # This assumes the two spectrum viewers have the same x-axis shape and
         # wavelength solution, which should always hold
-        if change['old'] is None:
-            pass
-        else:
-            name = change['name']
-            if name not in ['min', 'max']:
-                return
-            new_idx = int(np.around(change['new']))
-            spec1d = self.app.get_viewer('table-viewer')._selected_data["spectrum-viewer"]
-            extend_by = int(self.app.data_collection[spec1d]["World 0"].shape[0])
-            world = self._extend_world(spec1d, extend_by)
-
-            scales = self.app.get_viewer('spectrum-viewer').scales
-            old_val = getattr(scales['x'], name)
-
-            # Warn the user if they've panned far enough away from the data
-            # that the viewers will desync
-            try:
-                val = world[new_idx+extend_by]
-            except IndexError:
-                val = old_val
-                msg = "Warning: panning too far away from the data may desync \
-                       the 1D and 2D spectrum viewers"
-                msg = SnackbarMessage(msg, color='warning', sender=self)
-                self.app.hub.broadcast(msg)
-            if val != old_val:
-                setattr(scales['x'], name, val)
+        table_viewer = self.app.get_viewer('table-viewer')
+
+        if self._update_in_progress or table_viewer.row_selection_in_progress:
+            return
+
+        min_2d = self._scales2d.min
+        max_2d = self._scales2d.max
+
+        spec1d = table_viewer._selected_data["spectrum-viewer"]
+        extend_by = int(self.app.data_collection[spec1d]["World 0"].shape[0])
+        world = self._extend_world(spec1d, extend_by)
+
+        idx_min = int(np.around(min_2d)) + extend_by
+        idx_max = int(np.around(max_2d)) + extend_by
+
+        # Warn the user if they've panned far enough away from the data
+        # that the viewers will desync
+        # Note: Because of the flipped data workaround, idx_min can be > idx_max
+        if not (0 <= idx_min < len(world) and 0 <= idx_max < len(world)):
```

Comment:
It bothers me a bit that `len(world)` is calculated twice in the same line, but I am being too nitpicky, perhaps...

Reply:
I don't mind changing that.